### ISSUE TITLE
Add support for nested classes

### ISSUE THREAD

**Issue by kiddkai** [mhallin/forest#3]
eg.

``` clj
(defstylesheet
  [.container { ... } ]
  [.show-when-container-hover { :display "none" }]
  [.container:hover .show-when-container-hover { :display "block" }])
```

``` clj
(defstylesheet
  [.container { ... } ]
  [.show-when-container-hover { :display "none" }]
  [.container:hover>.show-when-container-hover { :display "block" }])
```

Currently have error in both cases.

Thanks :smile:

**Comment by mhallin** [comment]
Yeah, we'll probably need some way of declaring combined selectors. Probably nest it again, something like:

``` clojure
(defstylesheet sample
  [[.container:hover > .show-when-container-hover] {:display "block"}])
```

...and maybe patching some of CSS behavior in the process, such as using `>` by default.

**Comment by ghost** [comment]
This might be related, but is it currently possible to use selectors on elements within a module? In normal stylus I would do something like:

```
.module
  header
    ...some styles only for headers under module
  p
    ...some styles only for p under module
```

Which leads me to believe I could do something like

```
(defstylesheet ss
  [.module {:display "flex"}]
  [:header {:color "blue"}] ;; doesn't work
```

But that `:header` is applied globally to all headers. 

Perhaps something like...

```
(defstylesheet ss
  [.module {:display "flex"}
    [:header {:color "blue"}]]
```

Is this supported and the syntax is not documented? Is this something css modules doesn't do at all?

If so, this would allow the markup to stay very lean without having to attach a `{:class some-selector}` to almost every element.

**Comment by ghost** [comment]
To further illustrate my point... the lack of nested selectors makes both `hiccup` templates and the `defstylsheet` a bit of a maintainability nightmare:

``` cljs
  [:div {:class styles/parameters}
   [:header {:class styles/header}
    [:h1 {:class styles/h1} "Parameters"]
    [:span.slash {:class styles/slash} "/"]
    [:h2 {:class styles/h2} "When and how does this discussion end?"]]
   [:div {:class styles/duration}
    [:h3 {:class styles/h3} "duration"]
    [:div {:class styles/input-group}]]]])
```

``` cljs
(defstylesheet styles
  [.parameters {
                :border-top "1px solid #ddd"
                :border-bottom "1px solid #ddd"
                :display "flex"
                :flex-direction "column"
                :padding (ems 1 0)
                }]
  [.header {
            :align-items "flex-end"
            :display "flex"
            }]
  [.h1 {:display "flex"
        :font-size (ems 2)}]
  [.slash {:color "#ddd"
           :font-size (ems 1.5)
           :padding (ems 0 (/ 1 1.5))}]
  [.h2 {
        :font-size (ems 1.5)
        }]
  [.duration {}]
  [.h3 {
        :font-size (ems 1.25)
        }]
  [.input-group {
                 :display "flex"
                 }]
  [.button {
            :display "flex"
            }]
  [.active {:composes button
            :background-color "#ddd"}])
```

There is just simply too much fluff added to make this a viable solution at the moment, in my opinion. A component that could be expressed in much fewer characters suddenly has an explosion of classes and excess markup. What's more, is that this is what I would consider a trivial, small component, and it will only get worse with larger components. Perhaps my issue is with the philosophy of css-modules, though, and not this implementation.

**Comment by mhallin** [comment]
The syntax for nesting that I'm currently trying out on the master branch looks like this:

``` clojure
(defstylesheet ss
  ;; For any descendant
  [(descendant .module header) {:color "blue"}]

  ;; For immediate descendants
  [(> .module header) {:color "blue"}]
)
```

For maintainability reasons, I'm trying to make the more scalable way of writing CSS the easiest one. While descendant selectors are convenient, they get pretty messy once you start nest modules.

There's a small conflict between `:composes`, descendant selectors, and the way this library works in that it can't be clever about how it exports class names. For example, compare these examples:

``` clojure
(defstylesheet ss
  [.base-class {:color "blue"}]
  [.subclass {:font-weight "bold" :composes base-class}])
```

``` css
/* Compiled roughly to */
.mangled__base-class {
    color: blue;
}

.mangled__subclass {
    font-weight: bold;
}

/* With the following Clojurescript bindings
   base-class => "mangled__base-class"
   subclass => "mangled__base-class mangled__subclass"
*/
```

vs.

``` clojure
(defstylesheet ss
  [.base-class {:color "blue"}]
  [(descendant .module .subclass) {:font-weight "bold" :composes base-class}])
```

``` css
/* Suggested resulting CSS */
.mangled__base-class, .mangled__module .mangled__subclass {
    color: blue;
}

.mangled__module .mangled__subclass {
    font-weight: bold;
}

/* With the following Clojurescript bindings:
   base-class => "mangled__base-class"
   module => "mangled__module"
   subclass => "mangled__subclass"
*/
```

So the second example can't take advantage of the more slimmed CSS and "clever" class name export.

---

On a similar note, I'm not a big fan of the Less/Sass style nesting where you can do stuff like:

``` sass
.module {
    header {
        color: blue;
    }
}
```

Since it becomes very hard to read on larger codebases. The `&` sign also makes things even more confusing since you now have to keep track of the context in your head. Exceptions to this rule could be pseudo classes, pseudo elements, and _maybe_ media queries on the same element since they are usually grouped together anyway.

I should also mention that this library is pulling inspiration from https://github.com/noprompt/garden. So if you haven't seen it already, it might be better suited for your needs if you want to succinctly express larger selectors or use nesting. It's more similar to Sass/Less than this library.

**Comment by kiddkai** [comment]
:+1: for make things simpler.

**Comment by ghost** [comment]
@mhallin I see your point. Thanks for taking time to consider my thoughts. I'm currently working on a macro (haven't quite gotten it right yet) that will allow me to get around having to use quoted strings, but instead can write something like `(. some-class {:display :flex :flex-direction :column})`. Doesn't seem like much, but I'm so used to regular css that I spend tens of minutes every day trying to figure out why `:align-items center` doesn't work :P.

**Comment by ghost** [comment]
Any word on the progress of this? Currently under development, waiting until you have more time, or just not being worked on? It's something that would prove very helpful with regards to overriding library-provided css files.. currently I don't see a way to, for example, enhance or tweak styles pulled in by a date-picker.css library.

**Comment by mhallin** [comment]
I am working on it in the moments I have some time left... There's support for compiling `(descendant a b ...)` and `(> a b ...)` in `defstylesheet` on the master branch right now, but it doesn't work well with the `:composes` keyword and there are no integration tests.

**Comment by mhallin** [comment]
Support for the `descendant` and `>` selectors have now been released in 0.2.0, but they can't interact with `:composes`. This of course limits their usefulness but might be a reasonable first implementation. The limitation is inherited from CSS Modules.

Does this cover your use case?

**Comment by kiddkai** [comment]
It looks good to me. Thanks
